### ISSUE TITLE
Including comments

### ISSUE THREAD

**Issue by 06kellyjac** [open-policy-agent/conftest#421]
Are there any plans to include comments in some/all input types?
I'm writing policy for Dockerfiles & I'd like to ignore policies based on comments.

Could also hold value for a policy that denies phrases within comments etc.

**Comment by Blokje5** [comment]
@06kellyjac instead of comments we now support exceptions: https://www.conftest.dev/exceptions/. Exceptions provide a mechanism to exclude certain input from certain rules. Hopefully that covers your use case.

**Comment by 06kellyjac** [comment]
I'm using exceptions for global ignores but I want to exempt an individual line

Fail on unpinned FROMs
```Dockerfile
# ignore lack of tag
FROM something as builder
RUN echo "hey"

FROM otherthing:tag
RUN echo "hi"
```

or

Fail on unpinned apt dependencies
```Dockerfile
FROM something as base
# ignore unpinned version
RUN apt-get update && apt-get install foo

FROM otherthing:tag
RUN apt-get update && apt-get install bar==1.0.0
```

There are a couple workarounds I can do but it makes things clunkier for people writing the Dockerfiles & I'd rather just look at the comments.

Also what if I wanted to enforce there are no `TODO` comments, etc

**Comment by Blokje5** [comment]
It would be nice if we could support such a use case, although I am not sure if it is feasible. There are three requirements here:

1. The input type supports comments. For example JSON, which Conftest supports as an input type does not support comments.

2. The parser we use has a mechanism for identifying comments and associating it with the next node. I am not sure if all parsers support this, although it is possible. I have not verified this yet.

3. We can identify the line that triggered the Rego policy and associate it with the relevant line in the input file. E.g. if there is a rule checking tag lines, this should return the line in the Dockerfile. This is harder then it sounds as we only get the results of a query from the Open Policy Agent libraries. And besides that we often need to restructure the data going into the policy engine.

Due to the inconsistencies between input types and the difficulty associating Rego policy queries with a specific line in the input file, I am not sure if this will be feasible in the short term.

**Comment by 06kellyjac** [comment]
I'm not necessarily suggesting comments for ignoring be a built in feature, I just want comments to show up in the parse

If an input type doesn't support comments then there wont be any comments :shrug: 

---

Docker example:
```Dockerfile
FROM golang:1.13 AS builder
WORKDIR /blah
# a comment
RUN do something 
CMD [ "http", "8080" ]
```

```JSON
[
	[
		{
			"Cmd": "from",
			"Flags": [],
			"JSON": false,
			"SubCmd": "",
			"Value": [
				"golang:1.13",
				"AS",
				"builder"
			]
		},
		{
			"Cmd": "workdir",
			"Flags": [],
			"JSON": false,
			"SubCmd": "",
			"Value": [
				"/blah"
			]
		},
        {
			"Cmd": "comment",
			"Flags": [],
			"JSON": false,
			"SubCmd": "",
			"Value": [
				"a comment"
			]
		},
        {
			"Cmd": "run",
			"Flags": [],
			"JSON": false,
			"SubCmd": "",
			"Value": [
				"do something"
			]
		},
		{
			"Cmd": "cmd",
			"Flags": [],
			"JSON": true,
			"SubCmd": "",
			"Value": [
				"http",
				"8080"
			]
		}
	]
]
```

Structure could be different



In the case of Dockerfiles I don't think there's any case where you can put comments on the same line e.g. `WORKDIR /something # comment`
But if it did it could look like so:

```JSON
[
    [
		{
			"Cmd": "workdir",
			"Flags": [],
			"JSON": false,
			"SubCmd": "",
			"Value": [
                "/something"
            ],
            "Comment": "a comment"
		}
    ]
]
```

being empty if there's no comment

```JSON
[
    [
		{
			"Cmd": "workdir",
			"Flags": [],
			"JSON": false,
			"SubCmd": "",
			"Value": [
                "/something"
            ],
            "Comment": ""
		}
    ]
]
```

The question then would be how to get the previous element in the array...

**Comment by 06kellyjac** [comment]
Ignoring the comments, the main issue I have is I want to ignore rules in some parts of the file while enforcing in others.
At the very least ignoring or enforcing across a whole `FROM` block
e.g. ignoring all unpinned apt versions in the `base` layer
https://github.com/open-policy-agent/conftest/issues/421#issuecomment-702063746

**Comment by jpreese** [comment]
@06kellyjac sorry for the delay, reading through some of the older issues. I think this should be possible today. Looking through your examples, and then the result of `conftest parse` today -- the Docker parser now has a `stage` field which exposes an incrementing value. Stage 0 being base, Stage 1 being the next etc:


```JSON
[
        [
                {
                        "Cmd": "from",      
                        "Flags": [],        
                        "JSON": false,      
                        "Stage": 0,
                        "SubCmd": "",       
                        "Value": [
                                "something",
                                "as",       
                                "base"      
                        ]
                },
                {
                        "Cmd": "run",       
                        "Flags": [],        
                        "JSON": false,
                        "Stage": 0,
                        "SubCmd": "",
                        "Value": [
                                "apt-get update \u0026\u0026 apt-get install foo"
                        ]
                },
                {
                        "Cmd": "from",
                        "Flags": [],
                        "JSON": false,
                        "Stage": 1,
                        "SubCmd": "",
                        "Value": [
                                "otherthing:tag"
                        ]
                },
                {
                        "Cmd": "run",
                        "Flags": [],
                        "JSON": false,
                        "Stage": 1,
                        "SubCmd": "",
                        "Value": [
                                "apt-get update \u0026\u0026 apt-get install bar==1.0.0"
                        ]
                }
        ]
]
```

With that you would write a policy that changes based on the stage value.

**Comment by 06kellyjac** [comment]
The stage field will be helpful for targeting a specific stage to ignore but if I want the comments to hold ignore rules like linters such as hadolint, shellcheck, eslint, markdown-lint etc

```
FROM something
# hadolint ignore=DL3003,SC1035
RUN cd /tmp && echo "hello!"
RUN other stuff
```

If the conftest team is still anti including comments in the parse results I'll need to find an alternative method

**Comment by jpreese** [comment]
> The stage field will be helpful for targeting a specific stage to ignore but if I want the comments to hold ignore rules like linters such as hadolint, shellcheck, eslint, markdown-lint etc
> 
> ```
> FROM something
> # hadolint ignore=DL3003,SC1035
> RUN cd /tmp && echo "hello!"
> RUN other stuff
> ```
> 
> If the conftest team is still anti including comments in the parse results I'll need to find an alternative method

Not opposed to the idea, just from previous comments it seemed like you were trying to ignore unpinned versions from the base layer which would now be possible with the stage field.

**Comment by jpreese** [comment]
@06kellyjac looks like they _do_ expose a `PrevComments` slice which we can piggyback off of.

Given:
```dockerfile
FROM openjdk:8-jdk-alpine
# This is a comment
VOLUME /tmp
ARG DEPENDENCY=target/dependency
COPY ${DEPENDENCY}/BOOT-INF/lib /app/lib
COPY ${DEPENDENCY}/META-INF /app/META-INF
COPY ${DEPENDENCY}/BOOT-INF/classes /app

RUN apk add --no-cache python3 python3-dev build-base && pip3 install awscli==1.18.1

ENTRYPOINT ["java","-cp","app:app/lib/*","hello.Application"]

FROM somethingelse:foo
RUN do the thing

# Another comment

COPY shenan igans
```


Results in:
```JSON
examples/docker/Dockerfile
[
        [
                {
                        "Cmd": "from",
                        "Flags": [],
                        "JSON": false,
                        "Stage": 0,
                        "SubCmd": "",
                        "Value": [
                                "openjdk:8-jdk-alpine"
                        ]
                },
                {
                        "Cmd": "comment",
                        "Flags": [],
                        "JSON": false,
                        "Stage": 0,
                        "SubCmd": "",
                        "Value": [
                                "This is a comment"
                        ]
                },
                {
                        "Cmd": "volume",
                        "Flags": [],
                        "JSON": false,
                        "Stage": 0,
                        "SubCmd": "",
                        "Value": [
                                "/tmp"
                        ]
                },
                {
                        "Cmd": "arg",
                        "Flags": [],
                        "JSON": false,
                        "Stage": 0,
                        "SubCmd": "",
                        "Value": [
                                "DEPENDENCY=target/dependency"
                        ]
                },
                {
                        "Cmd": "copy",
                        "Flags": [],
                        "JSON": false,
                        "Stage": 0,
                        "SubCmd": "",
                        "Value": [
                                "${DEPENDENCY}/BOOT-INF/lib",
                                "/app/lib"
                        ]
                },
                {
                        "Cmd": "copy",
                        "Flags": [],
                        "JSON": false,
                        "Stage": 0,
                        "SubCmd": "",
                        "Value": [
                                "${DEPENDENCY}/META-INF",
                                "/app/META-INF"
                        ]
                },
                {
                        "Cmd": "copy",
                        "Flags": [],
                        "JSON": false,
                        "Stage": 0,
                        "SubCmd": "",
                        "Value": [
                                "${DEPENDENCY}/BOOT-INF/classes",
                                "/app"
                        ]
                },
                {
                        "Cmd": "run",
                        "Flags": [],
                        "JSON": false,
                        "Stage": 0,
                        "SubCmd": "",
                        "Value": [
                                "apk add --no-cache python3 python3-dev build-base \u0026\u0026 pip3 install awscli==1.18.1"
                        ]
                },
                {
                        "Cmd": "entrypoint",
                        "Flags": [],
                        "JSON": true,
                        "Stage": 0,
                        "SubCmd": "",
                        "Value": [
                                "java",
                                "-cp",
                                "app:app/lib/*",
                                "hello.Application"
                        ]
                },
                {
                        "Cmd": "from",
                        "Flags": [],
                        "JSON": false,
                        "Stage": 1,
                        "SubCmd": "",
                        "Value": [
                                "somethingelse:foo"
                        ]
                },
                {
                        "Cmd": "run",
                        "Flags": [],
                        "JSON": false,
                        "Stage": 1,
                        "SubCmd": "",
                        "Value": [
                                "do the thing"
                        ]
                },
                {
                        "Cmd": "comment",
                        "Flags": [],
                        "JSON": false,
                        "Stage": 1,
                        "SubCmd": "",
                        "Value": [
                                "Another comment"
                        ]
                },
                {
                        "Cmd": "copy",
                        "Flags": [],
                        "JSON": false,
                        "Stage": 1,
                        "SubCmd": "",
                        "Value": [
                                "shenan",
                                "igans"
                        ]
                }
        ]
]
```

https://github.com/open-policy-agent/conftest/pull/534

**Comment by 06kellyjac** [comment]
Awesome! If you're all happy to add that it'd be great for the ignore thing I want to implement & some of the comment policies I've been thinking of.

What happens for comments on the same line? I guess I can just build the PR and give it a go

```
FROM openjdk:8-jdk-alpine
VOLUME /tmp # This is a comment
```

**Comment by jpreese** [comment]
> Awesome! If you're all happy to add that it'd be great for the ignore thing I want to implement & some of the comment policies I've been thinking of.
> 
> What happens for comments on the same line? I guess I can just build the PR and give it a go
> 
> ```
> FROM openjdk:8-jdk-alpine
> VOLUME /tmp # This is a comment
> ```

I don't believe that would be valid. To my knowledge comments must be on their own line.

**Comment by 06kellyjac** [comment]
Yeah you're right, I thought I'd double check

**Comment by jpreese** [comment]
Resolved via #534
